### DISCUSSION TITLE
DPY-2035: invalid object type name

### DISCUSSION THREAD

**Discussion by TheSteveBurgess** [oracle/python-oracledb#149] [Q&A] [answered]
Hi,

I'm trying to access a subtype in a custom package, but i receive a "DPY-2035: invalid object type name" error as well as a "OCI-22303" error from Oracle when trying to access it.  

I'm using the below code in Python

```connection.gettype("MYSUBTYPE")```

I can access other non subtypes fine in the same way.

I can also access the subtype fine through datagrip. 

Is there a specific way that i should be accessing subtypes?

Thanks,

Steve.

**Comment by anthony-tuininga** [comment]
Can you provide more detail? Specifically, the create type statements that result in this error? Thanks!

**Reply by TheSteveBurgess** [reply]
This is the declaration for the subtype in the custom package

```SUBTYPE TIMECARD_INFO IS APPS.HXC_SELF_SERVICE_TIME_DEPOSIT.TIMECARD_INFO;```

Is that what you're looking for?

**Reply by anthony-tuininga** [reply]
Is this a PL/SQL type?

**Reply by TheSteveBurgess** [reply]
Honestly I'm not entirely sure, my knowledge of Oracle is unfortunately pretty limited.

I believe APPS.HXC_SELF_SERVICE_TIME_DEPOSIT.TIMECARD_INFO is a user defined type as I'm having to use the thick client to access it in Python. I guess that makes TIMECARD_INFO a user defined type as it's a subtype of APPS.HXC_SELF_SERVICE_TIME_DEPOSIT.TIMECARD_INFO?

**Reply by anthony-tuininga** [reply]
Since it contains 3 segments I suspect it is a PL/SQL type and `APPS.HXC_SELF_SERVICE_TIME_DEPOSIT` is a PL/SQL package. So where is `TIMECARD_INFO` defined? Is that also inside a PL/SQL package?

**Reply by TheSteveBurgess** [reply]
That's in a custom package called xxx_time_pkg.

So I'm accessing it in Python with the below line of code

timecard_info_type = connection.gettype("APPS.XXX_TIME_PKG.TIMECARD_INFO")

**Reply by anthony-tuininga** [reply]
FYI, user-defined types work fine in thin mode, too (as of version 1.2.0). Can you try with that mode and see if you get a different error message? There was also a slight enhancement for handling PL/SQL package types in that version. What version are you using?

**Reply by TheSteveBurgess** [reply]
I upgraded to 1.2.2 a few days ago. Let me try it in thin mode and i'll come back with any errors.

Thanks!

Steve.

**Reply by TheSteveBurgess** [reply]
Errors at the same place but with a different error this time

```
 timecard_info_type = connection.gettype("APPS.XXX_TIME_PKG.TIMECARD_INFO")
  File "c:\users\*****\appdata\local\programs\python\python38\lib\site-packages\oracledb\connection.py", line 441, in gettype
    obj_type_impl = self._impl.get_type(self, name)
  File "src\oracledb\impl/thin/connection.pyx", line 361, in oracledb.thin_impl.ThinConnImpl.get_type
  File "src\oracledb\impl/thin/dbobject.pyx", line 906, in oracledb.thin_impl.ThinDbObjectTypeCache.get_type
  File "src\oracledb\impl/thin/dbobject.pyx", line 872, in oracledb.thin_impl.ThinDbObjectTypeCache._populate_type_info
  File "src\oracledb\impl/thin/dbobject.pyx", line 846, in oracledb.thin_impl.ThinDbObjectTypeCache._parse_tds
  File "src\oracledb\impl/thin/dbobject.pyx", line 794, in oracledb.thin_impl.ThinDbObjectTypeCache._parse_element_type
  File "src\oracledb\impl/thin/dbobject.pyx", line 683, in oracledb.thin_impl.ThinDbObjectTypeCache._determine_element_objtype
TypeError: 'NoneType' object is not iterable
```
Am i right in thinking to use thin mode i just dont call oracledb.init_oracle_client()?

**Reply by anthony-tuininga** [reply]
Yes, that is correct. Looks like you found a bug! Can you provide a test case that demonstrates the problem?

**Reply by TheSteveBurgess** [reply]
I'll try my best :)

**Reply by anthony-tuininga** [reply]
If you can provide the definition of the type on which the subtype is based, that might be sufficient for me to replicate.

**Comment by TheSteveBurgess** [comment]
Here's the definition of the package as a whole:

https://etrm.live/etrm-12.1.1/etrm.oracle.com/pls/et1211d9/etrm_pnav1e99-6.html?c_name=HXC_SELF_SERVICE_TIME_DEPOSIT&c_owner=APPS&c_type=PACKAGE&c_detail_type=source

timecard_info is defined on line 122. It looks like is a table of building_block_info, which is defined on line 97.  building_block_info looks like it takes most of it's types from hxc_time_building_blocks, the definition for that is here (i believe) - https://etrm.live/etrm-12.1.1/etrm.oracle.com/pls/et1211d9/etrm_pnavaf53.html

Hope thats what you need :)

**Comment by santoclemenzi** [comment]
Any news about this issue? Getting the exact same error message with oracledb 1.4.2

**Reply by cjbj** [reply]
@santoclemenzi do you have a (smaller) testcase that you can share?

**Reply by santoclemenzi** [reply]
Sure! I'll try to give you some examples of what I'm doing. This is the python code which is giving me the error:

```python
import oracledb

db_user = 'USER_SCHEMA'
db_pw = 'USER_PW'
db_dsn = 'myserver.mycompany.com/SERVICENAME'
connection = oracledb.connect(user=db_user, password=db_pw, dsn=db_dsn, encoding='UTF-8')
cur = connection.cursor()
po_ret = cur.var(int)
po_msg = cur.var(str)
type_str_verbali = connection.gettype('PKG11_STR.str_verbali')
```

I'm getting the error as soon as I hit the line `type_str_verbali = connection.gettype('PKG11_STR.str_verbali')`

This is the traceback:

```
Traceback (most recent call last):
  File "<pyshell#10>", line 1, in <module>
    type_str_verbali = connection.gettype('PKG11_STR.str_verbali')
  File "C:\Users\Clemenzi\AppData\Roaming\Python\Python39\site-packages\oracledb\connection.py", line 437, in gettype
    obj_type_impl = self._impl.get_type(self, name)
  File "src\\oracledb\\impl/thin/connection.pyx", line 377, in oracledb.thin_impl.ThinConnImpl.get_type
  File "src\\oracledb\\impl/thin/dbobject.pyx", line 924, in oracledb.thin_impl.ThinDbObjectTypeCache.get_type
  File "src\\oracledb\\impl/thin/dbobject.pyx", line 879, in oracledb.thin_impl.ThinDbObjectTypeCache._populate_type_info
  File "C:\Users\Clemenzi\AppData\Roaming\Python\Python39\site-packages\oracledb\errors.py", line 127, in _raise_err
    raise exc_type(_Error(message)) from cause
oracledb.exceptions.ProgrammingError: DPY-2035: invalid object type name: "PKG11_STR.str_verbali"
```

This is how the cutom type is defined:

```SQL
CREATE OR REPLACE PACKAGE ESSE3_UNIBOCCONI_PREPROD.pkg11_str IS
	//[...]
	TYPE str_verbali IS RECORD
    (
        lotto_id                      p11_verbali.lotto_id%TYPE,
        verb_id                       p11_verbali.verb_id%TYPE,
        cds_id                        p11_verbali.cds_id%TYPE,
        cds_cod                       p11_verbali.cds_cod%TYPE,
        cds                           p11_verbali.cds%TYPE,
        ad_id                         p11_verbali.ad_id%TYPE,
        ad_cod                        p11_verbali.ad_cod%TYPE,
        ad                            p11_verbali.ad%TYPE,
        data_esa                      p11_verbali.data_esa%TYPE,
        voto                          p11_verbali.voto%TYPE,
        causale                       p11_verbali.causale%TYPE,
        doc_pres_id                   p11_verbali.doc_pres_id%TYPE,
        cfu                           p11_verbali.cfu%TYPE,
        parz_flg                      p11_verbali.parz_flg%TYPE,
        parz_cfu                      p11_verbali.parz_cfu%TYPE,
        err_num                       p11_verbali.err_num%TYPE,
        img_path                      p11_verbali.img_path%TYPE,
        pres_stu                      p11_verbali.pres_stu%TYPE,
        batch_id                      p11_verbali.batch_id%TYPE,
        ud_cod                        p11_verbali.ud_cod%TYPE,
        img_id                        p11_verbali.img_id%TYPE,
        mat_stu_id                    p11_verbali.mat_stu_id%TYPE,
        mat_cds_id                    p04_mat.cds_id%TYPE,
        mat_aa_ord_id                 p04_mat.aa_ord_id%TYPE,
        mat_pds_id                    p04_mat.pds_id%TYPE,
        mat_id                        p11_verbali.mat_id%TYPE,
        stu_id                        p11_verbali.stu_id%TYPE,
        doc_1memb_id                  p11_verbali.doc_1memb_id%TYPE,
        doc_2memb_id                  p11_verbali.doc_2memb_id%TYPE,
        data_app                      p11_verbali.data_app%TYPE,
        app_id                        p11_verbali.app_id%TYPE,
        app_log_id                    p11_verbali.app_log_id%TYPE,
        ad_stu_cod                    p11_verbali.ad_stu_cod%TYPE,
        ad_stu                        p11_verbali.ad_stu%TYPE,
        ad_stu_id                     p11_verbali.ad_stu_id%TYPE,
        cds_stu_cod                   p11_verbali.cds_stu_cod%TYPE,
        cds_stu                       p11_verbali.cds_stu%TYPE,
        cds_stu_id                    p11_verbali.cds_stu_id%TYPE,
        aa_ses_id                     p11_verbali.aa_ses_id%TYPE,
        nome                          p11_verbali.nome%TYPE,
        cognome                       p11_verbali.cognome%TYPE,
        matricola                     p11_verbali.mat%TYPE,
        verb_num                      p11_verbali.verb_num%TYPE,
        stato                         p11_verbali.stato%TYPE,
        stato_warn                    p11_verbali.stato_warn%TYPE,
        warn_num                      p11_verbali.warn_num%TYPE,
        ses_id                        p11_verbali.ses_id%TYPE,
        ses_cod                       p11_verbali.ses_cod%TYPE,
        tipo_gest_app_cod             p11_lotti_verb.tipo_gest_app_cod%TYPE,
        tipo_verb_cod                 p11_verbali.tipo_verb_cod%TYPE,
        carica_prova_flg              tipi_verbali.carica_prova_flg%TYPE,
        annulla_verbale_coll_flg      tipi_verbali.annulla_verbale_coll_flg%TYPE,
        cod_livello_uscita_lingua     p11_verbali.cod_livello_uscita_lingua%TYPE,
        aa_off_id                     p11_verbali.aa_off_id%TYPE,
        registrazione_verbali_flg     PLS_INTEGER,
        tipo_svolgimento_esame_cod    p11_verbali.tipo_svolgimento_esame_cod%TYPE
    );
	//[...]
```
All %TYPE within the type are a mix of VARCHAR2, NUMBER and DATE

My environment:
- Windows 10 21H2
- Python 3.9.13
- oracledb 1.4.2

I'm not completely sure what Oracle client the library is using, this is my path env var:
![image](https://github.com/oracle/python-oracledb/assets/22657674/6ca548cb-8c0c-4b25-876f-0d5e93d6bc59)

Thank you!

**Reply by santoclemenzi** [reply]
Oracle version is "Oracle Database 19c Enterprise Edition Release 19.0.0.0.0"

**Reply by anthony-tuininga** [reply, marked as answer]
> I'm not completely sure what Oracle client the library is using

You can find out by getting the result of this:

```
oracledb.clientversion()
```

> I'm getting the error as soon as I hit the line type_str_verbali = connection.gettype('PKG11_STR.str_verbali')

This may be as simple as converting the name to all uppercase. Type names are case sensitive. If you create the type without quotes, however, Oracle Database converts the name to all uppercase!

**Comment by santoclemenzi** [comment]
> This may be as simple as converting the name to all uppercase. Type names are case sensitive. If you create the type without quotes, however, Oracle Database converts the name to all uppercase!

That did the trick! Converting the name to all uppercase solved the issue, thank you very much for your help!
